### PR TITLE
[WIP] Fix Markers size with perspective

### PR DESCRIPTION
**Pull request by kevinyamauchi**
In this PR, I aim to fix the rendered Markers size when viewing with a perspective camera and `scaling` is True (https://github.com/napari/napari/issues/2637). I think the issue is that the way the points scaled did not account for the scaling due to perspective. My understanding is that when `scaling` is `True`, the Markers should have a consistent size in the visual data coordinates and when `scaling` is `False`, the markers should have a constant size in canvas pixels.

Currently, the vertex shader branches based on if scaling is True/False. I would like to change this so that the code gets substituted to remove this unnecessary branch, but I was having trouble getting the `$` substitution working

### CODE REVIEW

File: vispy/visuals/markers.py [hunk 1]
```diff
@@ -30,8 +29,31 @@
...
+        vec4 shifted_vert_dir = (shifted_vert - vec4(a_position, 1.0)) / 100;
+    
+        vec4 lr_vec = normalize(shifted_vert_dir);
+    
+        vec4 radius = $viewtransformf(vec4(a_position, 1.0) + lr_vec * a_size);
+        radius = radius/radius.w - vert_on_doc/vert_on_doc.w;
+        float radius_px = length(radius);
+        float size_scale = radius_px / a_size;
+        $v_size = size_scale * a_size;
+        v_edgewidth = size_scale * a_edgewidth;
```
Comment (brisvag) [inline vispy/visuals/markers.py:50]: I find this a bit less confusing:
```glsl
float radius_px = length(radius);
$v_size = radius_px;
v_edgewidth = a_edgewidth * radius_px / a_size;
```

**Review by brisvag — COMMENTED** [review]
I tried to reproduce the "wobbliness" I see in #2204 at low fovs, and though it looks different I think it's still happening here. Try to change the `point _cloud` example to have points of size `30` and `view.camera.fov = 0.2`... Not sure how related these things are though.

**Comment by kevinyamauchi** [comment]
> I tried to reproduce the "wobbliness" I see in #2204 at low fovs, and though it looks different I think it's still happening here. Try to change the point _cloud example to have points of size 30 and view.camera.fov = 0.2... Not sure how related these things are though.

Thanks for trying that! I just tried your example in #2204 and I can also see the "wobble". I'm not sure what the issue is. I'll try to dig into it later this week.

**Comment by brisvag** [comment]
As mentioned in #2204, there are some numerical instabilities concerning this approach. You should probably use the same trick I used there (with the `big_float`) to avoid them!

File: vispy/visuals/markers.py [hunk 2]
```diff
@@ -30,8 +29,31 @@
...
-    $v_size = a_size * u_px_scale * u_scale;
-    v_edgewidth = a_edgewidth * float(u_px_scale);
+
+    if (u_scale < 0) {
+        // left-right vector in the canvas
+        vec4 vert_on_doc = $viewtransformf(vec4(a_position, 1.0));
+        vec4 shifted_vert_on_doc = vert_on_doc + vec4(100, 0, 0, 0);
+    
+        vec4 shifted_vert = $viewtransformi(shifted_vert_on_doc);
+        shifted_vert = (shifted_vert / shifted_vert.w);
```
Comment (brisvag) [inline vispy/visuals/markers.py:39]: Are you sure about this `/ shifted_vert.w`? As far as I can tell, it should not be here. (See wobblyness comments in #2204). You already divide by `w` when calculating the radius in px later.

**Comment by djhoese** [comment]
@brisvag Should this be closed based on your #2209 PR?

**Comment by kevinyamauchi** [comment]
Yes, I think this should be superceded by #2209!

**Pull request closed without merging**